src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -4321,6 +4321,107 @@ void C2_MacroAssembler::vector_maskall_operation(KRegister dst, Register src, in
   }
 }
 
+
+//
+// Following is lookup table based popcount computation algorithm:-
+//       Index   Bit set count
+//     [ 0000 ->   0,
+//       0001 ->   1,
+//       0010 ->   1,
+//       0011 ->   2,
+//       0100 ->   1,
+//       0101 ->   2,
+//       0110 ->   2,
+//       0111 ->   3,
+//       1000 ->   1,
+//       1001 ->   2,
+//       1010 ->   3,
+//       1011 ->   3,
+//       1100 ->   2,
+//       1101 ->   3,
+//       1111 ->   4 ]
+//  a. Count the number of 1s in 4 LSB bits of each byte. These bits are used as
+//     shuffle indices for lookup table access.
+//  b. Right shift each byte of vector lane by 4 positions.
+//  c. Count the number of 1s in 4 MSB bits each byte. These bits are used as
+//     shuffle indices for lookup table access.
+//  d. Add the bitset count of upper and lower 4 bits of each byte.
+//  e. Unpack double words to quad words and compute sum of absolute difference of bitset
+//     count of all the bytes of a quadword.
+//  f. Perform step e. for upper 128bit vector lane.
+//  g. Pack the bitset count of quadwords back to double word.
+//  h. Unpacking and packing operations as not needed for 64bit vector lane.
+void C2_MacroAssembler::vector_popcount_int(XMMRegister dst, XMMRegister src, XMMRegister xtmp1,
+                                            XMMRegister xtmp2, XMMRegister xtmp3, Register rtmp,
+                                            int vec_enc) {
+  if (VM_Version::supports_avx512_vpopcntdq()) {
+    vpopcntd(dst, src, vec_enc);
+  } else if (vec_enc == Assembler::AVX_512bit) {
+    assert(VM_Version::supports_avx512vlbw(), "");
+    movl(rtmp, 0x0F0F0F0F);
+    evpbroadcastd(xtmp1, rtmp, vec_enc);
+    evmovdqul(xtmp2, k0, ExternalAddress(StubRoutines::x86::vector_popcount_lut()), true, vec_enc, rtmp);
+    evpandd(xtmp3, k0, src, xtmp1, true, vec_enc);
+    vpshufb(xtmp3, xtmp2, xtmp3, vec_enc);
+    Assembler::evpsrlw(dst, k0, src, 4, true , vec_enc);
+    evpandd(dst, k0, dst, xtmp1, true, vec_enc);
+    vpshufb(dst, xtmp2, dst, vec_enc);
+    evpaddb(xtmp3, k0, dst, xtmp3, true, vec_enc);
+    evpxord(xtmp1, k0, xtmp1, xtmp1, true, vec_enc);
+    evpunpckhdq(dst, k0, xtmp3, xtmp1, true, vec_enc);
+    evpsadbw(dst, k0, dst, xtmp1, true, vec_enc);
+    evpunpckldq(xtmp2, k0, xtmp3, xtmp1, true, vec_enc);
+    evpsadbw(xtmp2, k0, xtmp2, xtmp1, true, vec_enc);
+    vpackuswb(dst, xtmp2, dst, vec_enc);
```

Comment:
This doesn't look correct for say 512 bit vector length. At this point, the xtmp2 has 64-bit popcount results for lower 8 integers and dst has 64-bit popcount results for upper 8 integers. The vpackuswb does interleaving between xtmp2 and dst at 128 bit lanes so the result is not in correct order.

Reply:
original vector of integer = [ a3  a2  a1  a0]
unpackldq =  [0, a1, 0, a0]
unpackhdq = [ 0, a3 , 0 , a2]
perform sum of absolute difference and store the result into LSB 16 bits of each quad word.
packuswb  packs at lane granularity i.e, 128 bits packed.
      128 bit               128 bit
[0, sa3, 0, sa2]  [ 0, sa1, 0, sa0 ]  =>   [ sa3, sa2 , sa1, sa0]

Reply:
The problem is at 512 bit level.

Reply:
```
original 512 bit vector holding 16 integers = [ a15 a14 a13 a12, a11 a10 a9 a8 ,a7 a6 a5 a4 ,a3 a2 a1 a0 ]
 unpackdq lower:
                512 bit (vec1) 
   128           128        128      128
 0 a13 0 a12  0 a9 0 a8  0 a5 0 a4  0 a1 0 a0
 unpackdq higher:
                512 bit (vec1) 
   128           128        128      128
 0 a15 0 a14  0 a11 0 a10  0 a7 0 a6  0 a3 0 a2
 Next sum of absolute difference operation followed by pack will squeez each 128 bit lane of two participant vectors
 and interleave them in resulatant vector.
         VEC1_L3     VEC1_L2     VEC1_L1   VEC1_L0              VEC2_L3     VEC2_L2      VEC2_L1    VEC2_L0           
 [ 0 sa13 0 sa12  0 sa9 0 sa8  0 sa5 0 sa4  0 sa1 0 sa0 ] [ 0 sa15 0 sa14  0 sa11 0 sa10  0 sa7 0 sa6  0 sa3 0 sa2 ]
 [ sa15 sa14 sa13 sa12   sa11 sa10 sa9 sa8  sa7 sa6 sa5 sa4  sa3 sa2 sa1 sa0]
```